lib/galaxy/di.py

```diff
@@ -0,0 +1,37 @@
+"""Dependency injection framework for Galaxy-type apps."""
+from typing import Optional, Type, TypeVar
+
+from lagom import Container as LagomContainer
+from lagom.exceptions import UnresolvableType
+
+T = TypeVar("T")
+
+
+class Container(LagomContainer):
+    """Abstraction around lagom to provide an dependency injection context.
+
+    Abstractions used by Galaxy should come through this interface so we can swap
+    out the backend as needed. For instance https://punq.readthedocs.io/en/latest/
+    containers look very nice and would allow us to also inject by name (e.g. for
+    config variables for instance).
+    """
+
+    def _register_singleton(self, dep_type: Type[T], instance: Optional[T] = None) -> T:
```

Comment:
This doesn't seem an internal method to me?

Reply:
It is only ever used by subclasses in their constructors or methods called by their constructors I think.

```python
lib/galaxy/app.py:        self._register_singleton(BasicApp, self)
lib/galaxy/app.py:        self._register_singleton(StructuredApp, self)
lib/galaxy/app.py:        self.config: Any = self._register_singleton(config.Configuration, config.Configuration(**kwargs))
lib/galaxy/app.py:        self.execution_timer_factory = self._register_singleton(ExecutionTimerFactory, ExecutionTimerFactory(self.config))
lib/galaxy/app.py:        self.application_stack = self._register_singleton(ApplicationStack, application_stack_instance(app=self))
lib/galaxy/app.py:        self.queue_worker = self._register_singleton(GalaxyQueueWorker, GalaxyQueueWorker(self))
lib/galaxy/app.py:        self._register_singleton(IdEncodingHelper, self.security)
lib/galaxy/app.py:        self._register_singleton(SharedModelMapping, self.model)
lib/galaxy/app.py:        self._register_singleton(GalaxyModelMapping, self.model)
lib/galaxy/app.py:        self._register_singleton(scoped_session, self.model.context)
lib/galaxy/app.py:        self.tag_handler = self._register_singleton(GalaxyTagHandler)
lib/galaxy/app.py:        self.user_manager = self._register_singleton(UserManager)
lib/galaxy/app.py:        self._register_singleton(GalaxySessionManager)
lib/galaxy/app.py:        self.hda_manager = self._register_singleton(HDAManager)
lib/galaxy/app.py:        self.history_manager = self._register_singleton(HistoryManager)
lib/galaxy/app.py:        self.job_search = self._register_singleton(JobSearch)
lib/galaxy/app.py:        self.dataset_collections_service = self._register_singleton(DatasetCollectionManager)
lib/galaxy/app.py:        self.workflow_manager = self._register_singleton(WorkflowsManager)
lib/galaxy/app.py:        self.workflow_contents_manager = self._register_singleton(WorkflowContentsManager)
lib/galaxy/app.py:        self.dependency_resolvers_view = self._register_singleton(DependencyResolversView, DependencyResolversView(self))
lib/galaxy/app.py:        self.test_data_resolver = self._register_singleton(TestDataResolver, TestDataResolver(file_dirs=self.config.tool_test_data_directories))
lib/galaxy/app.py:        self.library_folder_manager = self._register_singleton(FolderManager)
lib/galaxy/app.py:        self.library_manager = self._register_singleton(LibraryManager)
lib/galaxy/app.py:        self.role_manager = self._register_singleton(RoleManager)
lib/galaxy/app.py:        self.dynamic_tool_manager = self._register_singleton(DynamicToolManager)
lib/galaxy/app.py:        self.api_keys_manager = self._register_singleton(ApiKeyManager)
lib/galaxy/app.py:        self.file_sources = self._register_singleton(ConfiguredFileSources, ConfiguredFileSources.from_app_config(self.config))
lib/galaxy/app.py:        self.genomes = self._register_singleton(Genomes)
lib/galaxy/app.py:        self.data_provider_registry = self._register_singleton(DataProviderRegistry)
lib/galaxy/app.py:        self.job_metrics = self._register_singleton(JobMetrics, JobMetrics(self.config.job_metrics_config_file, app=self))
lib/galaxy/app.py:        self.error_reports = self._register_singleton(ErrorReports, ErrorReports(self.config.error_report_file, app=self))
lib/galaxy/app.py:        self.job_config = self._register_singleton(jobs.JobConfiguration)
lib/galaxy/app.py:        self.tool_cache = self._register_singleton(ToolCache)
lib/galaxy/app.py:        self.tool_shed_repository_cache = self._register_singleton(ToolShedRepositoryCache)
lib/galaxy/app.py:        self.watchers = self._register_singleton(ConfigWatchers)
lib/galaxy/app.py:        self.installed_repository_manager = self._register_singleton(InstalledRepositoryManager, InstalledRepositoryManager(self))
lib/galaxy/app.py:        self._register_singleton(Registry, self.datatypes_registry)
lib/galaxy/app.py:        self.data_managers = self._register_singleton(DataManagers)
lib/galaxy/app.py:        self.update_repository_manager = self._register_singleton(UpdateRepositoryManager, UpdateRepositoryManager(self))
lib/galaxy/app.py:        self.visualizations_registry = self._register_singleton(VisualizationsRegistry, VisualizationsRegistry(
lib/galaxy/app.py:        self.webhooks_registry = self._register_singleton(WebhooksRegistry, WebhooksRegistry(self.config.webhooks_dir))
lib/galaxy/app.py:        self.quota_agent = self._register_singleton(QuotaAgent, get_quota_agent(self.config, self.model))
lib/galaxy/app.py:        self.auth_manager = self._register_singleton(auth.AuthManager, auth.AuthManager(self.config))
lib/galaxy/app.py:        self.job_manager = self._register_singleton(manager.JobManager)
lib/galaxy/app.py:        self.trs_proxy = self._register_singleton(TrsProxy, TrsProxy(self.config))
lib/tool_shed/webapp/app.py:        self._register_singleton(idencoding.IdEncodingHelper, self.security)
lib/tool_shed/webapp/app.py:        self._register_singleton(SharedModelMapping, model)
lib/tool_shed/webapp/app.py:        self._register_singleton(mapping.ToolShedModelMapping, model)
lib/tool_shed/webapp/app.py:        self._register_singleton(scoped_session, self.model.context)
lib/tool_shed/webapp/app.py:        self._register_singleton(UserManager, UserManager)
lib/tool_shed/webapp/app.py:        self.auth_manager = self._register_singleton(auth.AuthManager, auth.AuthManager(self.config))
lib/tool_shed/webapp/app.py:        self.citations_manager = self._register_singleton(CitationsManager, CitationsManager(self))
lib/tool_shed/webapp/app.py:        self.quota_agent = self._register_singleton(QuotaAgent, NoQuotaAgent())
```